catroid/src/test/java/org/catrobat/catroid/test/utiltests/FlashUtilTest.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.test.utiltests;
+
+import android.hardware.Camera;
+
+import org.catrobat.catroid.camera.CameraManager;
+import org.catrobat.catroid.utils.FlashUtil;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import static junit.framework.Assert.assertTrue;
+import static junit.framework.TestCase.assertFalse;
+
+import static org.mockito.Mockito.when;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({CameraManager.class, Camera.class})
+public class FlashUtilTest {
+
+	private CameraManager oldCameraManagerInstance;
+
+	@Before
+	public void setUp() {
+		oldCameraManagerInstance = CameraManager.getInstance();
+	}
+
+	@After
+	public void resetInstance() {
+		CameraManager.setInstance(oldCameraManagerInstance);
+	}
```

Comment:
Setup & cleanup methods are unnecessary here. If other tests rely on CameraManager having a specific state, then they need to make sure it is in that state in their own tests.

Reply:
NoNoNoNo!
Always reset everything your tests changed in any kind of persistant state.

this can have so many side effects, not because a follow up test could need the camera manager directly, but the camera manager has 122 usages across the project, and if any of those its getting really tricky to debug and locate the problem. (eg. the compute dialog in formula editor accesses cameraManager)

i just spent 2 hours last week with a bug like that. someone mocked stage listener in a test and didnt have a teardown, leading to another test 100 tests apart in the test suite to fail due to a null pointer exception.

Reply:
Good point, didn't think about it that way...
Sorry for misinformation then!